### ISSUE TITLE
Error using cosine loss

### ISSUE THREAD

**Issue by DWhettam** [guybuk/ANCOR#2]
Hi,

thanks for making this code available! I'm having an issue trying to use a cosine loss with cifar100. Using ['ce', 'cosine'] as my losses I'm running into an error on the following line: 

`z = nn.functional.normalize(z, dim=1)` 

because z is of shape [128], so dim =1 is not valid. Should it be dim=0 instead for everything here?

**Comment by guybuk** [comment]
Hello,
The `cosine` option in `loss_factory.py` is actually not supported; it's related to an experiment that didn't make the paper, and I should probably remove it to avoid confusion. Anyhow, I don't think it will produce the results you expect, even if you change the code to `dim=0`.

I might be able to help if you elaborate on what is it you're trying to achieve. Are you trying to replace the cross-entropy in ANCOR with a cosine variant a la BYOL?

**Comment by DWhettam** [comment]
Hi,
Thanks for the reply! I may have misunderstood the paper/code in that case. I'm trying to replicate the paper as a starting point - my understanding from the paper is that you have a cross-entropy loss for the course classes and an InfoNCE loss for the fine classes. Is that correct? The default arguments seemed to be doing cross-entropy for both the course and fine classes, so I changed to the cosine loss to try to re-create the CE, InfoNCE scenario in the paper. 

Thanks!

**Comment by guybuk** [comment]
You are correct: Regular Cross Entropy for coarse classes, InfoNCE on the angular-normalized features for the fine classes.

In practice, InfoNCE can be implemented using Cross Entropy (see for example [here](https://github.com/facebookresearch/moco/blob/master/main_moco.py)), which is why the default arguments have cross entropy for both coarse and fine classes.

Also, apologies for the delayed responses. I can't seem to get push-notifications from Github to work :)

Let me know if there's anything else or if I can close this issue. Thanks!

**Comment by DWhettam** [comment]
Ah, I see - I didn't realise that detail. Thanks for the response, I'll close the issue
